src/client/dfs/duns.c

```diff
@@ -581,69 +582,78 @@ duns_create_lustre_path(daos_handle_t poh, const char *path,
 int
 duns_create_path(daos_handle_t poh, const char *path, struct duns_attr_t *attrp)
 {
-	char			pool[37], cont[37];
-	char			oclass[10], type[10];
-	char			str[DUNS_MAX_XATTR_LEN];
-	int			len;
-	bool			try_multiple = true;
-	int			rc;
-	bool			backend_dfuse = false;
+	char		pool[37], cont[37];
+	char		oclass[10], type[10];
+	char		str[DUNS_MAX_XATTR_LEN];
+	int		len;
+	bool		try_multiple = true;
+	int		rc;
+	bool		backend_dfuse = false;
+	char		*part;
+	char		*spart;
+	int		dfd;
+	struct statfs	fs;
+	int		unlink_flag = 0;
 
 	if (path == NULL) {
 		D_ERROR("Invalid path\n");
 		return EINVAL;
 	}
 
+	D_STRNDUP(part, path, MAXPATHLEN);
+	if (part == 0)
+		return ENOMEM;
+
+	spart = dirname(part);
+	dfd = open(spart, O_PATH | O_DIRECTORY);
+	if (dfd == -1) {
+		rc = errno;
+		D_ERROR("Failed to open parent directory %s: %s\n",
+			spart, strerror(rc));
+		D_FREE(part);
+		return rc;
+	}
+
+	rc = fstatfs(dfd, &fs);
+	if (rc == -1) {
+		rc = errno;
+		D_ERROR("Failed to statfs dir %s: %s\n",
+			spart, strerror(rc));
+		goto err_close;
+	}
+
+	D_FREE(part);
+
+	D_STRNDUP(part, path, MAXPATHLEN);
+	spart = basename(part);
+
 	if (attrp->da_type == DAOS_PROP_CO_LAYOUT_HDF5) {
 		/** create a new file if HDF5 container */
 		int fd;
 
-		fd = open(path, O_CREAT | O_EXCL,
-			  S_IRUSR | S_IWUSR | S_IRGRP | S_IWGRP | S_IROTH);
+		fd = openat(dfd, spart, O_CREAT | O_EXCL,
```

Comment:
what is the point of this change? why do we need opendir, openat in this case?

Reply:
I was thinking it would make it less error prone, and not susceptible to rename of entries in the path however there's no setxattrat() so the full path is still used in places.  Looking at the code I'd need to open the file with openat() then call fsetxattr() to achieve what I was intending to.

Reply:
fgetattr() doesn't work on Directories, so it's not possible to implement this properly using openat() anyway, I could solve the same potential problems using chdir(), but that would only be safe if duns_create_path() was in a binary, but as it's in a shared library that's not safe for other reasons.